include/arch/x86/arch.h

```diff
@@ -530,12 +531,112 @@ extern FUNC_NORETURN void _SysFatalErrorHandler(unsigned int reason,
 
 
 #ifdef CONFIG_X86_STACK_PROTECTION
+extern struct task_state_segment _main_tss;
+
+#ifdef CONFIG_X86_USERSPACE
+void _x86_drop_to_user_mode(void);
+
+static inline u32_t _arch_syscall_invoke5(u32_t call_id, u32_t arg1, u32_t arg2,
+					  u32_t arg3, u32_t arg4, u32_t arg5)
+{
+	u32_t ret;
+
+	__asm__ volatile("int %[vector]"
+			 : "=a" (ret)
+			 : "a" (call_id), "d" (arg1), "c" (arg2),
+			   "b" (arg3), "D" (arg4), "S" (arg5),
+			   [vector] "i" (CONFIG_X86_SYSCALL_VECTOR));
```

Comment:
Style: so, symbolic constants are always good, but not a whole lot of people understand the GCC syntax for named replacement arguments or even the fact that the arguments can be things other than registers.

I honestly think this would be a lot more readable and perfectly maintainable as "int 0x80", honestly.

Reply:
this function also marshals all the arguments in the correct registers. please look at the generated assembly. a bare "int 0x80" won't work, and I'm not comfortable with hard-coding the vector id.
if people don't understand GCC extended inline ASM syntax, they can google it.

Reply:
Of course "int 0x80" would work, that's all this is doing.  That directive says replace the named expression "vector" with the integer-typed expression CONFIG_X86_SYSCALL_VECTOR.  Then it hands that expanded string to the assembler which sees "int 0x80" (or more likely "int 128" I guess).

Obviously not a big deal.  I just don't really see this as all that readable given something (the syscall interrupt vector) that is a static "known to everyone and part of the well-documented calling conventions" expression.

Reply:
@andyross no, it's not just doing "int 0x80"

Please review https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html#Machine-Constraints

The arguments have to all go in very specific registers. It happens to be (by design) that call_id, arg1, arg2 will already be in the right registers for the IAMCU calling convention, but for SYS V they have to be moved, as well as any args past arg2.

Reply:
I'm not explaining myself well enough:  I totally get what you're doing with the **register** constraints here and it makes sense.  I'm just saying that you're jumping through hoops to insert "0x80" as a soft-configured string using a slightly obscure GCC feature and a ton of boilerplate, when realistically all you need to do is document "0x80 is the syscall trap" somewhere and use a 0x80 literal instead of the [vector] and Kconfig stuff.

Reply:
i think the point you are actually trying to argue is that "we don't need to allow the end user to configure what vector to use for system calls", is that a fair assessment?

Reply:
i think my thought process was that the IDT might be configured to have less than 256 entries, but if we're dealing with 4K pages I can't imagine people would be keen on optimizing the IDT size

Reply:
Exactly that.  Or if you really want to, do the parametrized INT instruction with a preprocessor##paste or some other generator and not the assembly trick.